### PR TITLE
Add GitHub actions

### PR DESCRIPTION
**Pull request by bradyoo**

### CODE REVIEW

**Comment by bradyoo** [comment]
jenkins, test this please.

**Comment by bradyoo** [comment]
alluxio-bot, test this please.

**Comment by bradyoo** [comment]
@gpang Want to make this official?

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [inline .github/workflows/checkstyle.yml:12]: where are these secret strings set?

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [inline .github/workflows/checkstyle.yml:46]: nit: split each env var into its own line for readability and same throughout similar files

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [inline .github/workflows/checkstyle.yml:58]: newline at eof and many other new files

File: .github/workflows/integrationtests.yml
Comment (Xenorith) [inline .github/workflows/integrationtests.yml:1]: rename to file integration_tests.yml or whatever is allowed as long as there's some kind of separation between words?

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [inline .github/workflows/checkstyle.yml:9]: are wildcards allowed for branches? we will need these to run on release branches (ex. `branch-2.x`) as well

File: .github/workflows/unittests.yml
Comment (Xenorith) [inline .github/workflows/unittests.yml:1]: ditto for file name here

File: dev/github/run_tests.sh
Comment (Xenorith) [inline dev/github/run_tests.sh:29]: this is a copy of the bash code in `run_checks.sh`; can they be consolidated?

**Review by Xenorith — CHANGES_REQUESTED** [review]
can you leave a README in `dev/github/` explaining the new layout of files and describe how github actions invokes the build? it's not very straightforward even as someone who knows how this currently works with jenkins

File: dev/github/run_tests.sh
Comment (bradyoo) [reply dev/github/run_tests.sh:29]: I don't intend to consolidate them. It's very very annoying to consolidate shell code in a reasonable way

Comment (bradyoo) [reply dev/github/run_tests.sh:29]: though I can probably delete a lot of things here so gonna do that first.

**Review by Xenorith — COMMENTED** [review]
it seems to me that you also removed the logic for "run only the necessary checks/tests based on the PR diff". can each build do its own diff and exits quickly if the diff is irrelevant to what the build is testing for?

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [inline .github/workflows/checkstyle.yml:9]: i was a bit lax in my previous comment. we need to run PR builds for the release branches but only starting from 2.4. i presume none of this would work for 2.3 or earlier, unless you think the full set of changes to make this work can be backported

**Comment by bradyoo** [comment]
@Xenorith diffing was really never really a thing.

File: .github/workflows/checkstyle.yml
Comment (bradyoo) [reply .github/workflows/checkstyle.yml:9]: Since the yml files don't exist in the branches themselves, it won't actually run them. For example, note that this code is in the github_actions branch in alluxio/alluxio but it's not actually running in master branch for others right now!

Comment (Xenorith) [reply .github/workflows/checkstyle.yml:9]: good point. so why is this line here at all?

**Comment by bradyoo** [comment]
alluxio-bot, test this please.

**Comment by bradyoo** [comment]
@gpang I got through the review from @Xenorith, would you like to review?

File: .github/workflows/checkstyle.yml
Comment (Xenorith) [reply .github/workflows/checkstyle.yml:9]: ping

**Review by Xenorith — CHANGES_REQUESTED** [review]
and readme on how github actions interacts with these files?

File: dev/github/run_tests.sh
Comment (gpang) [inline dev/github/run_tests.sh:34]: why are checkstyle and findbugs disabled in the PR builds?

**Review by gpang — COMMENTED** [review]
@bradyoo Thanks, I left a small question.

File: dev/github/run_tests.sh
Comment (bradyoo) [reply dev/github/run_tests.sh:34]: It is not it is under run_checks.sh checkstyle.yml in .github/ goes through a separate codepath that does builds, checkstyles, etc. note that you can see that category separately in the checks below.

Comment (bradyoo) [reply dev/github/run_tests.sh:34]: Checkstyle, Findbugs, Doc Check, etc. / build (pull_request)

File: .github/workflows/checkstyle.yml
Comment (bradyoo) [reply .github/workflows/checkstyle.yml:9]: because the line is still required for it to run. Even with this code being checked in, without this line of branch-2.* it will not run on it.

File: dev/github/run_tests.sh
Comment (gpang) [reply dev/github/run_tests.sh:34]: Ahhh, I see. Thanks!

File: dev/github/helper.sh
Comment (gpang) [inline dev/github/helper.sh:38]: This is for the checkstyle and findbugs task, right? This is also skipping checkstyle, findbugs, etc. Is this correct? I am a bit confused.

This is an output for this PR's checkstyle check. https://pipelines.actions.githubusercontent.com/110HvfICxw4TXSOnsd72jk32naeMsNgZF0xPcAxH5PwqdiBHUN/_apis/pipelines/1/runs/106/signedlogcontent/3?urlExpires=2020-09-30T17%3A18%3A37.2939398Z&urlSigningMethod=HMACV1&urlSignature=y%2B13RULZz8azm5yA9KZIWAnIzZTiBsg0JVWduMCxaDQ%3D 

However, in that output, I see
```
2020-09-30T08:20:56.6182722Z + mvn -Duser.home=/home/jenkins -T 4C clean install -Pdeveloper -Dfindbugs.skip -Dcheckstyle.skip -DskipTests -Dmaven.javadoc.skip -Dlicense.skip -Dsurefire.forkCount=2 -fn -DfailIfNoTests=false --fail-at-end
```

This looks like all the checks are skipped?

**Review by gpang — COMMENTED** [review]
@bradyoo I had additional questions about the checkstyle checks

File: dev/github/helper.sh
Comment (bradyoo) [reply dev/github/helper.sh]: Oh wow you are right. I accidentally dropped it while doing the refactor during the review process. I ended up undoing the refactor and moved helper.sh into the respective callers and modified the findbug/checkstyle skip parts.

**Comment by bradyoo** [comment]
alluxio-bot, test this please

File: dev/github/helper.sh
Comment (bradyoo) [reply dev/github/helper.sh]: Fixed now.

Comment (gpang) [reply dev/github/helper.sh]: cool, thanks! The mvn command looks correct now.

**Comment by bradyoo** [comment]
Tests passed.

**Review by gpang — APPROVED** [review]
LGTM

Thanks for this!

**Review by Xenorith — APPROVED** [review]

**Comment by bradyoo** [comment]
Feel free to merge. @gpang

**Comment by Xenorith** [comment]
actually i think we need to coordinate when this merges along with other associated changes, so let's hold off for a bit @gpang

**Comment by bradyoo** [comment]
Added retries into the github_action itself so that users don't have to do much retries.

**Comment by Xenorith** [comment]
alluxio-bot, merge this please

**Pull request merged**
